python/pyarrow/_compute.pyx

```diff
@@ -2182,3 +2255,175 @@ cdef CExpression _bind(Expression filter, Schema schema) except *:
+
+
+def register_function(func_name, num_args, function_doc, in_types,
+                      out_type, callback):
+    """
+    Register a user-defined-function
+
+    Parameters
+    ----------
+
+    func_name : str
+        function name 
+    num_args : int
+       number of arguments in the function
+    function_doc : dict
+        a dictionary object with keys 
+        ("summary", 
+        "description", 
+        "arg_names"
+        )
+    in_types : List[InputType]
+        list of InputType objects which defines the input 
+        types for the function
+    out_type : DataType
+        output type of the function
+    callback : callable
+        user defined function
+        function includes arguments equal to the number
+        of input_types defined. The return type of the 
+        function is of the type defined as output_type. 
+        The output is a datum object which can be
+        an Array or a ChunkedArray or a Table or a RecordBatch.
+
+    Example
+    -------
+
+    >>> from pyarrow import compute as pc
+    >>> from pyarrow.compute import register_function
+    >>> from pyarrow.compute import InputType
+    >>> 
+    >>> func_doc = {}
+    >>> func_doc["summary"] = "simple udf"
+    >>> func_doc["description"] = "add a constant to a scalar"
+    >>> func_doc["arg_names"] = ["x"]
+    >>> 
+    >>> def add_constant(array):
+    ...     return pc.call_function("add", [array, 1])
+    ... 
+    >>> 
+    >>> func_name = "py_add_func"
+    >>> arity = 1
+    >>> in_types = [InputType.array(pa.int64())]
+    >>> out_type = pa.int64()
+    >>> register_function(func_name, arity, func_doc,
+    ...                   in_types, out_type, add_constant)
+    >>> 
+    >>> func = pc.get_function(func_name)
+    >>> func.name
+    'py_add_func'
+    >>> ans = pc.call_function(func_name, [pa.array([20])])
+    >>> ans
+    <pyarrow.lib.Int64Array object at 0x10c22e700>
+    [
+    21
+    ]
+    """
+    cdef:
+        c_string c_func_name
+        CArity c_arity
+        CFunctionDoc c_func_doc
+        CInputType in_tmp
+        vector[CInputType] c_in_types
+        PyObject* c_callback
+        shared_ptr[CDataType] c_type
+        COutputType* c_out_type
+        CScalarUdfBuilder* c_sc_builder
+        CStatus st
+        CScalarUdfOptions* c_options
+        object obj
+
+    if func_name and isinstance(func_name, str):
+        c_func_name = tobytes(func_name)
+    else:
+        raise ValueError("func_name should be str")
+
+    if num_args and isinstance(num_args, int):
+        assert num_args > 0
+        if num_args == 0:
+            c_arity = CArity.Nullary()
+        elif num_args == 1:
+            c_arity = CArity.Unary()
+        elif num_args == 2:
+            c_arity = CArity.Binary()
+        elif num_args == 3:
+            c_arity = CArity.Ternary()
+        elif num_args > 3:
+            c_arity = CArity.VarArgs(num_args)
+    else:
+        raise ValueError("arity must be an instance of Arity")
```

Comment:
In any case, even if we were to check, these would be TypeError, and the messages reference the wrong types.